File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        def reverse(node):
            prev = None
            curr = temp = node
            
            while curr:
                temp = curr.next
                curr.next = prev
                prev = curr
                curr = temp
            return prev
            
        carry = 0
        tot = 0
        curSum = 0
        temp = res = ListNode()
        
        while l1 and l2:
            curSum = l1.val + l2.val + carry
            tot = curSum % 10
            carry = int(curSum/10)
            temp.next = ListNode(tot, temp.next)
            l1 = l1.next
            l2 = l2.next
        
        while l1:
            curSum = l1.val + carry
            tot = curSum % 10
            carry = int(curSum/10)
            temp.next = ListNode(tot, temp.next)
            l1 = l1.next
        
        while l2:
            curSum = l2.val + carry
            tot = curSum % 10
            carry = int(curSum/10)
            temp.next = ListNode(tot, temp.next)
            l2 = l2.next
        
        if carry:
            temp.next = ListNode(carry, temp.next)
        
        return reverse(res.next)
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
#         carry=0
#         tot=0
#         def reverseLL(head):
#             if not head or not head.next:
#                 return  head
#             temp=prev=None
#             curr=head
#             while curr:
#                 temp=curr.next
#                 curr.next=prev
#                 prev=curr
#                 curr=temp
#             return prev
#         if not l1:
#             return l2
#         elif not l2:
#             return l1
#         #l1=reverseLL(l1)
#         #l2=reverseLL(l2)
#         dummy=ListNode()
#         temp=dummy
#         while l1 and l2:
#             a=l1.val
#             b=l2.val
#             tot=(a+b+carry)%10
#             carry=(a+b+carry)//10
#             temp.next=ListNode(tot,temp.next)
#             l1=l1.next
#             l2=l2.next
#         if l1:
#             while l1:
#                 tot=(l1.val+carry)%10
#                 carry=(l1.val+carry)//10
#                 temp.next=ListNode(tot,temp.next)
#                 l1=l1.next
                
#         elif l2:
#             while l2:
#                 tot=(l2.val+carry)%10
#                 carry=(l2.val+carry)//10
#                 temp.next=ListNode(tot,temp.next)
#                 l2=l2.next
#         if carry>0:
#             temp.next=ListNode(carry,temp.next)
#         p=dummy.next
#         p=reverseLL(p)
#         return p   

File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int total = 0, carry = 0;
        ListNode dummy = new ListNode(0);
        ListNode temp = dummy;
        
        while ((l1 != null) || (l2 != null)){
            total = carry;
            if (l1 != null){
                total += l1.val;
                l1 = l1.next;
            }
            if (l2 != null){
                total += l2.val;
                l2 = l2.next;
            }
            carry = total / 10;
            total = total % 10;
            temp.next = new ListNode(total);
            temp = temp.next;
        }
        if (carry != 0){
            temp.next = new ListNode(carry);
        }
        return dummy.next;    
    
        
    }
}